packages/protect/src/lib/protect.ts

```diff
@@ -0,0 +1,261 @@
+  }
+
+  function getProtectType(step: FRStep): ProtectType {
+    const cbs = step.getCallbacksOfType(CallbackType.MetadataCallback);
+
+    if (!cbs.length) {
+      return 'none';
+    }
+
+    const protectMetadataCb = cbs.find((cb) => {
+      const metadataCallback = cb as MetadataCallback;
+      const data = metadataCallback.getData() as { _type: string; _action: string };
+      return data._type === 'PingOneProtect';
+    });
+
+    if (!protectMetadataCb) {
+      return 'none';
+    }
+
+    const data = (protectMetadataCb as MetadataCallback).getData() as ProtectInitializeConfig;
+
+    return data._action === 'protect_initialize' ? 'initialize' : 'evaluate';
+  }
+
+  function setNodeClientError(step: FRStep, value: string): void {
+    // Check for native callback first
+    const nativeEvaluationCallback = step.getCallbacksOfType<PingOneProtectEvaluationCallback>(
+      CallbackType.PingOneProtectEvaluationCallback,
+    );
+    const nativeInitializeCallback = step.getCallbacksOfType<PingOneProtectInitializeCallback>(
+      CallbackType.PingOneProtectInitializeCallback,
+    );
+    const arr = [...nativeEvaluationCallback, ...nativeInitializeCallback];
+
+    if (arr.length) {
+      const cb = arr[0];
+      cb.setClientError(value);
+      return;
+    }
+
+    // If we are here, we are dealing with Marketplace nodes
+    const cbs = step.getCallbacksOfType<HiddenValueCallback>(CallbackType.HiddenValueCallback);
+
+    if (!cbs.length) {
+      return;
+    }
+
+    const clientErrorCb = cbs.find((hiddenValueCallback) => {
+      const output = hiddenValueCallback.getOutputByName<string>('id', '');
+      return output === 'clientError';
+    });
+
+    if (!clientErrorCb) {
+      return;
+    }
+
+    clientErrorCb.setInputValue(value);
+  }
+
+  function setNodeInputValue(step: FRStep, value: string): void {
+    // Check for native callback first
+    try {
+      const nativeCallback = step.getCallbackOfType<PingOneProtectEvaluationCallback>(
+        CallbackType.PingOneProtectEvaluationCallback,
+      );
+
+      nativeCallback?.setData(value);
+      return;
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    } catch (err) {
+      // Do nothing
+    }
+
+    // If we are here, we are dealing with Marketplace nodes
+    const cbs = step.getCallbacksOfType<HiddenValueCallback>(CallbackType.HiddenValueCallback);
+
+    if (!cbs.length) {
+      return;
+    }
+
+    const inputCb = cbs.find((hiddenValueCallback) => {
+      const output = hiddenValueCallback.getOutputByName<string>('id', '');
+      return output === 'pingone_risk_evaluation_signals';
+    });
+
+    if (!inputCb) {
+      return;
+    }
+
+    inputCb.setInputValue(value);
+  }
+
+  const protectAPI: Protect = {
+    start,
+    getData,
+    pauseBehavioralData,
+    resumeBehavioralData,
+    getPauseBehavioralData,
+    getNodeConfig,
+    getProtectType,
+    setNodeClientError,
+    setNodeInputValue,
+  };
```

Comment:
I'd like to avoid this pattern where we have a big function that returns an object that isn't apparent until the very end of the parent function body. Rather, can we just return an object at the top, and inline all of these functions.

In other words:

```
// rather than this
export async function protect() {
  // ... lots of function definitions
  const api = {
    // lots of assignment
  };
  return api;
}
```

```
// I'd rather see this
export async function protect() {
  return {
    // inline functions
  };
}
```